zerver/lib/import_realm.py

```diff
@@ -444,6 +452,53 @@ def fix_message_rendered_content(
             )
 
 
+def valid_near_link_parts(near_link_parts: list[str]) -> bool:
+    return (
+        3 <= len(near_link_parts) <= 7
+        and near_link_parts[0] == "#narrow"
+        and near_link_parts[1] in ["channel", "stream"]
+    )
+
+
+def fix_near_links_in_messages(rendered_content: str) -> str:
+    soup = BeautifulSoup(rendered_content, "html.parser")
+    near_links = soup.find_all(
+        lambda tag: tag.name == "a"
+        and tag.has_attr("href")
+        and ("#narrow/stream/" in tag.get("href") or "#narrow/channel/" in tag.get("href"))
+    )
+
+    if near_links == []:
+        return rendered_content
+
+    message_id_map = ID_MAP["message"]
+    channel_id_map = ID_MAP["stream"]
+    for link in near_links:
+        href = link["href"]
+        # This is necessary because only some near-link start
+        # with "/".
+        near_link_parts = " ".join(href.split("/")).split()
+
+        if not valid_near_link_parts(near_link_parts):
+            continue
+        if near_link_parts[1] == "stream":
+            near_link_parts[1] = "channel"
+        encoded_channel_parts = near_link_parts[2].split("-")
+        old_channel_id = int(encoded_channel_parts[0])
+        new_channel_id = channel_id_map.get(old_channel_id)
+        if new_channel_id is None:
+            continue
+        encoded_channel_parts[0] = str(new_channel_id)
+        near_link_parts[2] = "-".join(encoded_channel_parts)
+
+        if len(near_link_parts) == 7:
+            old_message_id = int(near_link_parts[6])
+            new_message_id = message_id_map.get(old_message_id)
+            near_link_parts[6] = str(new_message_id)
+        link["href"] = "/".join(near_link_parts)
```

Comment:
This logic very much makes assumptions about the order of operators in a URL. I think that could be OK for now, but let's make the core logic for making the decision here be a pure function about URLs that lives in `zerver/lib/url_decoding.py`, so that a future sweep to harden the logic can handle that well.

Also you likely want to ensure that what is returned definitely starts with a `/`.

Further I think you have a bug for `near` links to locations in DM conversations.

Reply:
Hmm, I'm still not sure how much of this logic I should move to `url_decoding`. I've moved a good chunk of the validation and link parsing/reassembling work to c71262df91762b029257e325c414dcadfea4d191 as a helper class called `NearLinkHandler`, but the logic to remap the object IDs in the link fragment is still in `fix_near_links_in_messages` in `import_realm.py`. 

Should those remapping logic (now `remap_near_link_message_id` & `remap_near_link_recipient_encoding`) be moved to `url_decoding` / `url_encoding` too? I didn't move them because I don't think they'd be reusable elsewhere..

Reply:
I think the remapping is likely not usable elsewhere, so you've probably got the right division of responsibility.